### PR TITLE
Shortcode based shop page for unsupported themes

### PR DESCRIPTION
**Pull request by mikejolley**
This is for #17550 

- If using an unsupported theme, rather than load archive-shop, the shop page is left as-is and the the content is filtered to insert shortcodes.
- Enhanced product shortcode to support pagination and disable caching.
- Added rows/cols options in settings to control default layout. 3 cols works better across themes and this ensures pages have equal products per row.
- Made query class and filter widgets rely less on globals.
- Made some templates work without globals.

Note; taxonomies are left alone. I could not find a good way to make those use page.php so I’ve left it out of this PR and probably for good.

### CODE REVIEW

**Comment by claudiulodro** [comment]
@mikejolley should this be merged into a feature branch or are we just going for it?

File: includes/class-wc-query.php [hunk 1]
```diff
@@ -443,19 +428,21 @@ public function get_catalog_ordering_args( $orderby = '', $order = '' ) {
 		$args    = array(
```
Comment (claudiulodro) [inline includes/class-wc-query.php]: `Warning: strtolower() expects parameter 1 to be string, array given in /srv/www/wordpress-default/public_html/wp-content/plugins/woocommerce/includes/class-wc-query.php on line 426` when sorting shop by average rating.

Comment (claudiulodro) [reply includes/class-wc-query.php]: `WordPress database error: [Unknown column 'wp_postmeta.meta_value' in 'order clause']
SELECT wp_posts.ID FROM wp_posts WHERE 1=1 AND wp_posts.post_type = 'post' AND ((wp_posts.post_status = 'publish')) ORDER BY wp_postmeta.meta_value+0 DESC, wp_posts.post_date DESC LIMIT 0, 5` in the sidebar for some reason when sorting by popularity

File: includes/class-wc-post-types.php [hunk 2]
```diff
@@ -255,6 +255,9 @@ public static function register_post_types() {
 			$supports[] = 'comments';
 		}
 
+		$shop_page_id = wc_get_page_id( 'shop' );
+		$has_archive  = current_theme_supports( 'woocommerce' ) ? $shop_page_id && get_post( $shop_page_id ) ? urldecode( get_page_uri( $shop_page_id ) ) : 'shop' : false;
```
Comment (claudiulodro) [inline includes/class-wc-post-types.php]: This is a fun 1-liner.

**Comment by claudiulodro** [comment]
screen shot 2017-11-09 at 1 24 48 pm
----------
screen shot 2017-11-09 at 1 25 06 pm

We can fix it in a later PR, but the columns change isn't working well with the twenty* themes (Tested with twentytwelve and twentyseventeen). The products don't rescale properly with the columns settings, so they don't look correct at any setting except 4 columns.

**Comment by claudiulodro** [comment]
https://travis-ci.org/woocommerce/woocommerce/jobs/299786435

Unit tests need to get updated.

**Comment by claudiulodro** [comment]
Code looks good and this works great on all the themes I tried. 👍 Just needs a few things.

**Comment by mikejolley** [comment]
@claudiulodro I'll rebase on feature branch.

**Comment by claudiulodro** [comment]
Looks good. All issues have been fixed.

**Pull request merged**